{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differential expression testing human ICM\n",
    "For the human ICM data, we are interesting in one primary test, differences in ICM vs NF by cell type. To perform differential expression testing, we will use the summation + limma-voom approach, as proposed by Lun and Marioni (https://academic.oup.com/biostatistics/article/18/3/451/2970368?login=true) and shown to be slightly conservative, if anything, based on Zimmerman, Espeland, and Langefeld (https://www.nature.com/articles/s41467-021-21038-1).\n",
    "\n",
    "The model we will employ is `expr ~ 1 + disease + sex + age`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "plt.style.use('seaborn-white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/unix/chaffin/.local/lib/python3.7/site-packages/rpy2/robjects/pandas2ri.py:15: FutureWarning: pandas.core.index is deprecated and will be removed in a future version.  The public classes are available in the top-level namespace.\n",
      "  from pandas.core.index import Index as PandasIndex\n"
     ]
    }
   ],
   "source": [
    "# Libraries\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import os\n",
    "import scanpy as sc\n",
    "\n",
    "# Using rpy2 to call R code from python\n",
    "import rpy2\n",
    "import rpy2.situation as rsetup\n",
    "import rpy2.robjects as ro\n",
    "from rpy2.robjects.packages import importr\n",
    "from rpy2.robjects import r, pandas2ri\n",
    "from rpy2.robjects.conversion import localconverter\n",
    "\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# R packages required\n",
    "base = importr('base')\n",
    "edgeR = importr('edgeR')\n",
    "DESeq2 = importr('DESeq2')\n",
    "limma = importr('limma')\n",
    "stats = importr('stats')\n",
    "grdevices = importr('grDevices')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy: 1.7.2\n",
      "pandas: 1.2.4\n",
      "numpy: 1.18.1\n",
      "rpy2: 3.0.5\n"
     ]
    }
   ],
   "source": [
    "# Python package versions\n",
    "print('scanpy: ' + sc.__version__)\n",
    "print('pandas: ' + pd.__version__)\n",
    "print('numpy: ' + np.__version__)\n",
    "print('rpy2: ' + rpy2.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R version: 3.6.0\n",
      "edgeR: 3.30.3\n",
      "DESeq2: 1.20.0\n",
      "limma: 3.44.3\n",
      "stats: 3.6.0\n",
      "grdevices: 3.6.0\n"
     ]
    }
   ],
   "source": [
    "# R package version\n",
    "print('R version: ' + base.__version__)\n",
    "print('edgeR: ' + edgeR.__version__)\n",
    "print('DESeq2: ' + DESeq2.__version__)\n",
    "print('limma: ' + limma.__version__)\n",
    "print('stats: ' + stats.__version__)\n",
    "print('grdevices: ' + grdevices.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Turn off some annoying FutureWarnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set path for aggregation data and output data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_aggregation = '../../data/aggregation'\n",
    "path_diffexpress = '../../results/diffexpress'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create the folder to store results in if it doesn't exist\n",
    "if not os.path.exists(path_diffexpress):\n",
    "    os.mkdir(path_diffexpress)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparison: ICM vs NF\n",
    "For each cell type, we want to test for DE between ICM and NF across all genes. We will only sum counts across all nuclei of a given type in a sample if the sample has more than 25 nuclei of the type. Otherwise the sample will be excluded from the differential expression test. As a sensitivity check, and to ensure our results are robust, we will test for DE using both the CellRanger counts and the CellBender remove-background adjusted counts.\n",
    "\n",
    "We will filter out genes in < 1% of nuclei from BOTH the ICM and NF groups as these are too lowly expressed to test, as well as an additional filter of `filterByExpr` as implemented in edgeR."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Your filename has more than two extensions: ['.Neighbor', '.UMAP', '.Freeze1', '.h5ad'].\n",
      "Only considering the two last: ['.Freeze1', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.Neighbor', '.UMAP', '.Freeze1', '.h5ad'].\n",
      "Only considering the two last: ['.Freeze1', '.h5ad'].\n"
     ]
    }
   ],
   "source": [
    "# Read in the snRNAseq map\n",
    "adata = sc.read(os.path.join(path_aggregation,'humanICM.PostQC.Map/PostQC.Neighbor.UMAP.Freeze1.h5ad'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>disease</th>\n",
       "      <th>ICM</th>\n",
       "      <th>NF</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>leiden0.5</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "disease    ICM  NF\n",
       "leiden0.5         \n",
       "0            7   8\n",
       "1            7   8\n",
       "2            7   8\n",
       "3            7   8\n",
       "4            7   8\n",
       "5            7   8\n",
       "6            7   8\n",
       "7            7   8\n",
       "8            7   7\n",
       "9            3   2\n",
       "10           5   8\n",
       "11           0   1\n",
       "12           0   4\n",
       "13           0   1\n",
       "14           1   0\n",
       "15           1   0"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Inspect the number of samples that will be included in a DE test for each cell type\n",
    "base = (adata.obs[['individual','disease',\n",
    "                   'leiden0.5','sample']].groupby(['individual','disease','leiden0.5']).count()).reset_index()\n",
    "base['good'] = base['sample'] > 25\n",
    "base[['leiden0.5','disease','good']].groupby(['leiden0.5','disease']).sum().reset_index().pivot(index='leiden0.5',\n",
    "                                                                                                columns='disease',\n",
    "                                                                                                values='good')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will only be able to test for differential expression for clusters 0 - 10, as the remaining clusters only have enough nuclei from one condition. We will have to be cautious with cluster 9 (adipocytes) given only 2 samples are used from the NF group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group the leiden clusters into the cell types we want to test\n",
    "groups = {'pseudo_bulk':[str(x) for x in range(len(adata.obs['leiden0.5'].unique()))], # this is ALL clusters combined\n",
    "          'cardiomyocyte':['0'],\n",
    "          'fibroblast':['1'],\n",
    "          'endothelial1':['2'],\n",
    "          'pericyte':['3'],\n",
    "          'macrophage':['4'],\n",
    "          'endothelial2':['5'],\n",
    "          'lymphocyte':['6'],\n",
    "          'vsmc':['7'],\n",
    "          'lymphatic_endothelial':['8'],\n",
    "          'adipocyte':['9'],\n",
    "          'neuronal':['10']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Function to perform differential expression test\n",
    "'''\n",
    "def run_differential_expression(bdata, aggr_var, min_cell, group_var, target_group, ref_group,\n",
    "                                covar, covar_dtype, cellranger, out):\n",
    "    '''\n",
    "    bdata: the subset of nuclei to include in the test\n",
    "    aggr_var: the .obs key to sum counts across (e.g., individual)\n",
    "    min_cell: the minimum cells to consider in a sample for aggregating\n",
    "    group_var: the .obs key for the grouping variable\n",
    "    target_group: the target group of the group_var to compare to ref_group\n",
    "    ref_group: the reference group of the DE test\n",
    "    covar: covariates to include in the DE test\n",
    "    covar_dtype: the datatype for the covariates\n",
    "    cellranger: True/False indicating whether to use .layers['cellranger_raw'] or not\n",
    "    out: output file prefix\n",
    "    '''\n",
    "    # Aggregate the counts by the aggr_var\n",
    "    # This generates the \"pseudo-bulk\" count matrix\n",
    "    tobind = [] # fill in each aggregation group, column-wise\n",
    "    if cellranger:\n",
    "        cellranger_key = 'cellranger'\n",
    "        for ind in bdata.obs[aggr_var].unique():\n",
    "            bdata_subset = bdata[bdata.obs[aggr_var]==ind].copy()\n",
    "            if bdata_subset.shape[0] > min_cell:\n",
    "                append_vect = pd.DataFrame(bdata_subset.layers['cellranger_raw'].sum(0)).transpose()\n",
    "                append_vect.columns = [ind]\n",
    "                tobind.append(append_vect) \n",
    "    else:\n",
    "        cellranger_key = 'cellbender'\n",
    "        for ind in bdata.obs[aggr_var].unique():\n",
    "            bdata_subset = bdata[bdata.obs[aggr_var]==ind].copy()\n",
    "            if bdata_subset.shape[0] > min_cell:\n",
    "                append_vect = pd.DataFrame(bdata_subset.X.sum(0)).transpose()\n",
    "                append_vect.columns = [ind]\n",
    "                tobind.append(append_vect)\n",
    "    summed_exp = pd.concat(tobind, axis = 1)\n",
    "    summed_exp.index = bdata.var.index\n",
    "    # Only keep samples that are in the groups being compared\n",
    "    keep_samp = bdata.obs[bdata.obs[group_var].isin([target_group, ref_group])][aggr_var].unique()\n",
    "    summed_exp = summed_exp[[x for x in keep_samp.tolist() if x in summed_exp.columns]]\n",
    "    \n",
    "    # Filter out lowly expressed genes -- this is always done on CellBender counts\n",
    "    # so that a similar set of genes goes into the DE test (even if using the CellRanger counts for testing)\n",
    "    # This removes genes in < 1% of both comparison groups\n",
    "    keep_gene = bdata.var[(np.array((bdata[bdata.obs[group_var] == target_group].X > 0).mean(0) > 0.01).squeeze())|\n",
    "                          (np.array((bdata[bdata.obs[group_var] == ref_group].X > 0).mean(0) > 0.01).squeeze())].index.tolist()\n",
    "    summed_exp = summed_exp[summed_exp.index.isin(keep_gene)]\n",
    "    print('\\tGenes with > 1% in either group = ' + str(summed_exp.shape[0]))\n",
    "\n",
    "    # Set up data in R\n",
    "    meta_data = bdata.obs[[aggr_var, group_var] + covar].drop_duplicates().reset_index().drop('index',1)\n",
    "    meta_data[aggr_var] = ['X' + str(x) if x[0].isdigit() else str(x) for x in meta_data[aggr_var]]\n",
    "\n",
    "    with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "        all_counts = ro.conversion.py2rpy(summed_exp) # count data\n",
    "    with localconverter(ro.default_converter + pandas2ri.converter): \n",
    "        meta_data_r = ro.conversion.py2rpy(meta_data) # meta data\n",
    "    ro.r.assign('all_counts', all_counts)\n",
    "    ro.r.assign('meta_data_r', meta_data_r)\n",
    "    \n",
    "    # Remove MT and RB genes\n",
    "    ro.r('mt_genes <- rownames(all_counts)[regexpr(\"MT-\",rownames(all_counts)) != -1]')\n",
    "    ro.r('rb_genes <- rownames(all_counts)[regexpr(\"^RP[SL]\",rownames(all_counts)) != -1]')\n",
    "    ro.r('all_counts <- all_counts[which(!rownames(all_counts) %in% mt_genes & !rownames(all_counts) %in% rb_genes),]')\n",
    "    \n",
    "    # Set up grouping variable and covariates\n",
    "    ro.r('group <- with(meta_data_r, ' + group_var + '[match(colnames(all_counts), ' + aggr_var + ')])')\n",
    "    for cov, covt in zip(covar, covar_dtype):\n",
    "        if covt == 'factor':\n",
    "            ro.r('as.factor(' + cov + ' <- with(meta_data_r, ' + cov + '[match(colnames(all_counts), ' + aggr_var + ')]))')\n",
    "        elif covt == 'numeric':\n",
    "            ro.r('as.numeric(' + cov + ' <- with(meta_data_r, ' + cov + '[match(colnames(all_counts), ' + aggr_var + ')]))')    \n",
    "\n",
    "    # filter out genes with low expression -- we'll use the EdgeR function to determine what \"low\" is\n",
    "    ro.r('keep.exprs <- filterByExpr(all_counts, group = group)')\n",
    "    ro.r('all_counts <- all_counts[keep.exprs, ]')\n",
    "    print(ro.r('paste0(\"      Genes retained after filterByExpr = \",dim(all_counts)[1])'))\n",
    "    print(ro.r('table(group)'))\n",
    "    \n",
    "    # Set up for limma-voom\n",
    "    ro.r('sample.formula <- ~ 1+group+' + '+'.join(covar))\n",
    "    ro.r('sample.data <- data.frame(group=factor(group, levels=c(\"' + ref_group + '\",\"' + target_group + '\")),' + ','.join(covar) + ')')\n",
    "    ro.r('design <- model.matrix(sample.formula, sample.data)')\n",
    "    # Normalize data\n",
    "    sf = DESeq2.estimateSizeFactorsForMatrix(ro.r('all_counts')) # DESeq2 normalization\n",
    "    ro.r.assign('sf', sf)\n",
    "    ro.r('eff.lib <- sf*mean(colSums(all_counts))')\n",
    "    print('\\tdata normalized...')\n",
    "    \n",
    "    # Run Voom\n",
    "    ro.r('y <- DGEList(all_counts, lib.size=eff.lib)')\n",
    "    grdevices.pdf(file=out + '_' + cellranger_key + '_voom.pdf')\n",
    "    ro.r('v.all <- voomWithQualityWeights(y, design, plot=T)')\n",
    "    grdevices.dev_off()\n",
    "    \n",
    "    ro.r('vfit <- lmFit(v.all, design)')\n",
    "    ro.r('vfit <- eBayes(vfit, robust=TRUE)')\n",
    "    ro.r('vres <- topTable(vfit, n=Inf, sort.by=\"P\", coef=\"group' + target_group +'\")')\n",
    "    ro.r('write.table(vres,\"' + out + '_' + cellranger_key + '_results.txt\", , sep=\"\\t\", quote=F)')\n",
    "\n",
    "    # Extra Voom diagnostic (plotSA)\n",
    "    grdevices.pdf(file=out + '_' + cellranger_key + '_plotSA.pdf')\n",
    "    ro.r('plotSA(vfit)')\n",
    "    grdevices.dev_off()\n",
    "    print('Done!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the folder to store these DE results\n",
    "if not os.path.exists(os.path.join(path_diffexpress, 'human_ICM_vs_NF')):\n",
    "    os.mkdir(os.path.join(path_diffexpress, 'human_ICM_vs_NF'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Testing celltype (CellBender) pseudo_bulk\n",
      "\tGenes with > 1% in either group = 12507\n",
      "[1] \"      Genes retained after filterByExpr = 12449\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellRanger) pseudo_bulk\n",
      "\tGenes with > 1% in either group = 12507\n",
      "[1] \"      Genes retained after filterByExpr = 12450\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellBender) cardiomyocyte\n",
      "\tGenes with > 1% in either group = 13467\n",
      "[1] \"      Genes retained after filterByExpr = 13236\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellRanger) cardiomyocyte\n",
      "\tGenes with > 1% in either group = 13467\n",
      "[1] \"      Genes retained after filterByExpr = 13381\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellBender) fibroblast\n",
      "\tGenes with > 1% in either group = 10704\n",
      "[1] \"      Genes retained after filterByExpr = 10267\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellRanger) fibroblast\n",
      "\tGenes with > 1% in either group = 10704\n",
      "[1] \"      Genes retained after filterByExpr = 10453\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellBender) endothelial1\n",
      "\tGenes with > 1% in either group = 8349\n",
      "[1] \"      Genes retained after filterByExpr = 6842\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellRanger) endothelial1\n",
      "\tGenes with > 1% in either group = 8349\n",
      "[1] \"      Genes retained after filterByExpr = 7851\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellBender) pericyte\n",
      "\tGenes with > 1% in either group = 7954\n",
      "[1] \"      Genes retained after filterByExpr = 6203\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellRanger) pericyte\n",
      "\tGenes with > 1% in either group = 7954\n",
      "[1] \"      Genes retained after filterByExpr = 7559\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellBender) macrophage\n",
      "\tGenes with > 1% in either group = 9415\n",
      "[1] \"      Genes retained after filterByExpr = 7061\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellRanger) macrophage\n",
      "\tGenes with > 1% in either group = 9415\n",
      "[1] \"      Genes retained after filterByExpr = 7829\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellBender) endothelial2\n",
      "\tGenes with > 1% in either group = 10504\n",
      "[1] \"      Genes retained after filterByExpr = 3765\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellRanger) endothelial2\n",
      "\tGenes with > 1% in either group = 10504\n",
      "[1] \"      Genes retained after filterByExpr = 4674\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellBender) lymphocyte\n",
      "\tGenes with > 1% in either group = 8890\n",
      "[1] \"      Genes retained after filterByExpr = 1566\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellRanger) lymphocyte\n",
      "\tGenes with > 1% in either group = 8890\n",
      "[1] \"      Genes retained after filterByExpr = 2331\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellBender) vsmc\n",
      "\tGenes with > 1% in either group = 10113\n",
      "[1] \"      Genes retained after filterByExpr = 1740\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellRanger) vsmc\n",
      "\tGenes with > 1% in either group = 10113\n",
      "[1] \"      Genes retained after filterByExpr = 2296\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellBender) lymphatic_endothelial\n",
      "\tGenes with > 1% in either group = 11053\n",
      "[1] \"      Genes retained after filterByExpr = 1847\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   7 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellRanger) lymphatic_endothelial\n",
      "\tGenes with > 1% in either group = 11053\n",
      "[1] \"      Genes retained after filterByExpr = 2122\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  7   7 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellBender) adipocyte\n",
      "\tGenes with > 1% in either group = 13893\n",
      "[1] \"      Genes retained after filterByExpr = 5571\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  3   2 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellRanger) adipocyte\n",
      "\tGenes with > 1% in either group = 13893\n",
      "[1] \"      Genes retained after filterByExpr = 5922\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  3   2 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellBender) neuronal\n",
      "\tGenes with > 1% in either group = 11151\n",
      "[1] \"      Genes retained after filterByExpr = 955\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  5   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n",
      "Testing celltype (CellRanger) neuronal\n",
      "\tGenes with > 1% in either group = 11151\n",
      "[1] \"      Genes retained after filterByExpr = 1308\"\n",
      "\n",
      "group\n",
      "ICM  NF \n",
      "  5   8 \n",
      "\n",
      "\tdata normalized...\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "# loop through each cell type and run the DE test\n",
    "# Once on cellbender counts and once on cellranger counts\n",
    "for celltype in groups.keys():\n",
    "    print('Testing celltype (CellBender) ' + celltype)\n",
    "    run_differential_expression(bdata = adata[adata.obs['leiden0.5'].isin(groups[celltype])].copy(),\n",
    "                                aggr_var = 'individual',\n",
    "                                min_cell = 25,\n",
    "                                group_var = 'disease',\n",
    "                                target_group = 'ICM',\n",
    "                                ref_group = 'NF',\n",
    "                                covar = ['sex', 'age'],\n",
    "                                covar_dtype = ['factor', 'numeric'],\n",
    "                                cellranger = False,\n",
    "                                out = os.path.join(path_diffexpress, 'human_ICM_vs_NF/' + celltype))\n",
    "    print('Testing celltype (CellRanger) ' + celltype)\n",
    "    run_differential_expression(bdata = adata[adata.obs['leiden0.5'].isin(groups[celltype])].copy(),\n",
    "                                aggr_var = 'individual',\n",
    "                                min_cell = 25,\n",
    "                                group_var = 'disease',\n",
    "                                target_group = 'ICM',\n",
    "                                ref_group = 'NF',\n",
    "                                covar = ['sex', 'age'],\n",
    "                                covar_dtype = ['factor', 'numeric'],\n",
    "                                cellranger = True,\n",
    "                                out = os.path.join(path_diffexpress, 'human_ICM_vs_NF/' + celltype))    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine all DE results into one results file and annotate additional information\n",
    "We want to aggregate all the information across all cell types into one file, and append in some additional information including:\n",
    "1. Pct > 0 in both ICM and NF groups\n",
    "2. Avg Expr in both ICM and NF groups\n",
    "3. Heuristic metric indiciating if the gene expression for a given gene is likely to come from background"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "function to calculate the background probability heuristic\n",
    "this is a metric that provides a sense of whether the gene\n",
    "expression is really driven by background contamination as\n",
    "CellBender is imperfect\n",
    "'''\n",
    "def calculate_bkg(adata, cluster_key, cluster):\n",
    "    # Approximate background profile across full dataset\n",
    "    # Based on cumulative distribution function\n",
    "    # Intuition: If a gene is not highly expressed in the entire experiment,\n",
    "    # it's not likely to be highly expressed in background\n",
    "    approx_background_profile = np.array(adata.X.sum(axis=0)).squeeze() / adata.X.sum()  # normalized\n",
    "    gene_order = np.argsort(approx_background_profile)\n",
    "    gene_to_cdf_x = np.argsort(gene_order)\n",
    "    cdf = np.cumsum(approx_background_profile[gene_order])\n",
    "    adata.var['gene_bkg_prob'] = cdf[[gene_to_cdf_x[i] for i in range(len(adata.var))]]\n",
    "    \n",
    "    # counts for inside and outside the target cluster\n",
    "    adata_subset = adata[adata.obs[cluster_key].isin(cluster)]\n",
    "    bc_out_group = list(set(adata.obs.index) - set(adata_subset.obs.index))\n",
    "    adata_out_group = adata[bc_out_group]\n",
    "    X_in = adata_subset.X\n",
    "    X_out = adata_out_group.X\n",
    "\n",
    "    # calculate fraction of cells in and out of group expressing > 0 counts\n",
    "    tp_0 = np.array((X_in > 0).sum(axis=0)).squeeze()\n",
    "    fp_0 = np.array((X_out > 0).sum(axis=0)).squeeze()\n",
    "    frac_exp_gr_0_target = tp_0 / adata_subset.shape[0]\n",
    "    frac_exp_gr_0_other = fp_0 / X_out.shape[0]\n",
    "\n",
    "    # calculate fraction of cells in and out of group expressing > 1 counts\n",
    "    # This can be helpful because there are often a lot of 1 counts for genes\n",
    "    tp_1 = np.array((X_in > 1).sum(axis=0)).squeeze()\n",
    "    fp_1 = np.array((X_out > 1).sum(axis=0)).squeeze()\n",
    "    frac_exp_gr_1_target = tp_1 / adata_subset.shape[0]\n",
    "    frac_exp_gr_1_other = fp_1 / X_out.shape[0]\n",
    "\n",
    "    # calculate standardized PPVs based on > 0 expression and > 1 expression\n",
    "    # this standardized PPV is equivalent to what would be calculated from\n",
    "    # specificity and sensivity with a set prevalence of 0.5 (e.g. assuming equal cluster size)\n",
    "    # Note: a small value is added to the denominator to avoid division by 0\n",
    "    ppv_0 = frac_exp_gr_0_target / (frac_exp_gr_0_target + frac_exp_gr_0_other + 1e-5)\n",
    "    ppv_1 = frac_exp_gr_1_target / (frac_exp_gr_1_target + frac_exp_gr_1_other + 1e-5)\n",
    "    \n",
    "    # take the average ppv0 and ppv1\n",
    "    mean_ppv = (ppv_0 + ppv_1) / 2.\n",
    "    prob_from_another_cluster = 1. - mean_ppv  # if very low mean_ppv, unlikely from this given cluster\n",
    "    bkg_prob = prob_from_another_cluster * adata.var['gene_bkg_prob'] # normalize this for the overall expression of the gene\n",
    "    return(bkg_prob)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Your filename has more than two extensions: ['.Neighbor', '.UMAP', '.Freeze1', '.h5ad'].\n",
      "Only considering the two last: ['.Freeze1', '.h5ad'].\n",
      "WARNING: Your filename has more than two extensions: ['.Neighbor', '.UMAP', '.Freeze1', '.h5ad'].\n",
      "Only considering the two last: ['.Freeze1', '.h5ad'].\n"
     ]
    }
   ],
   "source": [
    "# Read in the snRNAseq map\n",
    "adata = sc.read(os.path.join(path_aggregation,'humanICM.PostQC.Map/PostQC.Neighbor.UMAP.Freeze1.h5ad'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pseudo_bulk\n",
      "cardiomyocyte\n",
      "fibroblast\n",
      "endothelial1\n",
      "pericyte\n",
      "macrophage\n",
      "endothelial2\n",
      "lymphocyte\n",
      "vsmc\n",
      "lymphatic_endothelial\n",
      "adipocyte\n",
      "neuronal\n"
     ]
    }
   ],
   "source": [
    "# loop through each cell type and compile the information required\n",
    "target_group = 'ICM'\n",
    "ref_group = 'NF'\n",
    "group_var = 'disease'\n",
    "min_cell = 25 # matches above minimum cells required to aggregate\n",
    "tocat = []\n",
    "for gr in groups.keys():\n",
    "    print(gr)\n",
    "    # get the DE results\n",
    "    de_cb = pd.read_csv(os.path.join(path_diffexpress, 'human_ICM_vs_NF/' + gr + '_cellbender_results.txt'),\n",
    "                        sep='\\t').reset_index() # cellbender DE\n",
    "    de_cb.columns = ['index'] + [x + '_CB' for x in de_cb.columns[1:]] # rename columns\n",
    "    de_cr = pd.read_csv(os.path.join(path_diffexpress, 'human_ICM_vs_NF/' + gr + '_cellranger_results.txt'),\n",
    "                        sep='\\t').reset_index() # cellranger DE\n",
    "    de_cr.columns = ['index'] + [x + '_CR' for x in de_cr.columns[1:]] # rename columns\n",
    "    de = de_cb.merge(de_cr, on = 'index', how = 'left')\n",
    "    de = de.merge(adata.var.reset_index()[['index', 'gene_ids']], on = 'index', how = 'left')\n",
    "    # Add in the test-level annotation like number of cells/samples in the test per group\n",
    "    de['celltype'] = gr\n",
    "    de['grp1'] = target_group\n",
    "    de['grp2'] = ref_group\n",
    "    de['ncell_grp1'] = adata.obs[(adata.obs[group_var] == target_group)&(adata.obs['leiden0.5'].isin(groups[gr]))].shape[0]\n",
    "    de['ncell_grp2'] = adata.obs[(adata.obs[group_var] == ref_group)&(adata.obs['leiden0.5'].isin(groups[gr]))].shape[0]\n",
    "    de['nsamp_grp1'] = (adata.obs[(adata.obs[group_var] == target_group)&\n",
    "                                  (adata.obs['leiden0.5'].isin(groups[gr]))]['individual'].value_counts() > 25).sum()\n",
    "    de['nsamp_grp2'] = (adata.obs[(adata.obs[group_var] == ref_group)&\n",
    "                                  (adata.obs['leiden0.5'].isin(groups[gr]))]['individual'].value_counts() > 25).sum()\n",
    "        \n",
    "    # add the background prob, unless pseudo-bulk where not relevant\n",
    "    if gr == 'pseudo_bulk':\n",
    "        de['gene_bkg_prob'] = np.nan\n",
    "    else:\n",
    "        bkg_prob = pd.DataFrame(calculate_bkg(adata, 'leiden0.5', groups[gr])).reset_index()\n",
    "        bkg_prob.columns = ['gene','gene_bkg_prob']\n",
    "        de = de.merge(bkg_prob, left_on='index', right_on='gene', how='left')\n",
    "        \n",
    "    # pct and avg express for each gene\n",
    "    adata_clst = adata[adata.obs['leiden0.5'].isin(groups[gr])].copy()\n",
    "    sc.pp.normalize_total(adata_clst, target_sum=1e4)\n",
    "    sc.pp.log1p(adata_clst)\n",
    "    target_avg_expr = np.array(adata_clst[adata_clst.obs[group_var] == target_group].X.mean(0)).squeeze()\n",
    "    ref_avg_expr = np.array(adata_clst[adata_clst.obs[group_var] == ref_group].X.mean(0)).squeeze()\n",
    "    target_pct_gt0 = np.array((adata_clst[adata_clst.obs[group_var] == target_group].X > 0).mean(0)).squeeze()\n",
    "    ref_pct_gt0 = np.array((adata_clst[adata_clst.obs[group_var] == ref_group].X > 0).mean(0)).squeeze()    \n",
    "    expr_level = pd.DataFrame({'pctgt0_grp1':target_pct_gt0,\n",
    "                               'pctgt0_grp2':ref_pct_gt0,\n",
    "                               'avgexpr_grp1':target_avg_expr,\n",
    "                               'avgexpr_grp2':ref_avg_expr})\n",
    "    expr_level['index'] = adata.var.index.tolist()\n",
    "    de = de.merge(expr_level, left_on='index', right_on='index', how='left')\n",
    "    \n",
    "    # reorganize the data\n",
    "    de = de[['index', 'gene_ids', 'celltype', 'grp1', 'grp2', 'nsamp_grp1', 'nsamp_grp2',\n",
    "             'ncell_grp1', 'ncell_grp2', 'pctgt0_grp1', 'pctgt0_grp2', 'avgexpr_grp1', 'avgexpr_grp2',\n",
    "             'gene_bkg_prob',\n",
    "             'logFC_CB', 'AveExpr_CB', 't_CB', 'P.Value_CB', 'adj.P.Val_CB',\n",
    "             'logFC_CR', 'AveExpr_CR', 't_CR', 'P.Value_CR', 'adj.P.Val_CR']]\n",
    "    de.columns = ['gene', 'ensembl_id', 'celltype', 'grp1', 'grp2', 'nsamp_grp1', 'nsamp_grp2',\n",
    "                      'ncell_grp1', 'ncell_grp2', 'pctgt0_grp1', 'pctgt0_grp2', 'avgexpr_grp1', 'avgexpr_grp2',\n",
    "                      'gene_bkg_prob', 'logFC_CB', 'AveExpr_CB', 't_CB', 'P.Value_CB', 'adj.P.Val_CB',\n",
    "                     'logFC_CR', 'AveExpr_CR', 't_CR', 'P.Value_CR', 'adj.P.Val_CR']\n",
    "    tocat.append(de)\n",
    "\n",
    "# combine all the DE results across cell types together\n",
    "allres = pd.concat(tocat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "allres.to_csv(os.path.join(path_diffexpress, 'human_ICM_vs_NF_differential_expression_v2.0.txt.gz'),\n",
    "               sep='\\t', compression='gzip', na_rep='NA', index=None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
