{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 1: Aggregate all QC metrics per sample\n",
    "As the first step to constructing our snRNA-seq maps, we will aggregate all non-empty droplets from CellBender with appropriate QC metrics included. This requires 3 inputs per sample:\n",
    "1. CellRanger 4.0.0 count output: `<sampleID>.raw_feature_bc_matrix.h5`\n",
    "2. CellBender remove-background FPR = 0.01 h5: `<sampleID>_out_FPR_0.01_filtered.h5`\n",
    "3. scR-Invex output called per sample, collapsed to retain the number of exonic, intronic, and junction reads per barcode: `<sampleID>.scrinvex.collapsed.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "plt.style.use('seaborn-white')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Libraries\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import scanpy as sc\n",
    "import scrublet as scr\n",
    "import ndd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy: 1.7.2\n",
      "pandas: 1.2.4\n",
      "numpy: 1.18.1\n",
      "ndd: 1.6.3\n"
     ]
    }
   ],
   "source": [
    "print('scanpy: ' + sc.__version__)\n",
    "print('pandas: ' + pd.__version__)\n",
    "print('numpy: ' + np.__version__)\n",
    "print('ndd: ' + ndd.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Supplemental code to process the raw scR-Invex output\n",
    "Note: this was run on the direct output of scrinvex run on each CellRanger BAM file separately. It simply collapses the number of reads into a count of intronic-only, exonic-only, and overlapping intronic & exonic for each cell barcode."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    df = pd.read_csv('scrinvex_output.tsv', sep='\\t')\n",
    "    grp = df[['barcode','introns','junctions','exons']].groupby('barcode')\n",
    "    out = grp.sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set paths for aggregation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to CellRanger 4.0.0 output files\n",
    "# Files should be saved as <sampleID>.raw_feature_bc_matrix.h5\n",
    "path_cellranger = '../../data/CellRanger4.0.0'\n",
    "\n",
    "# Path to CellBender v2.1-alpha\n",
    "# Docker available at: us.gcr.io/broad-dsde-methods/cellbender:v2.1-alpha\n",
    "# Files should be saved as <sampleID>_out_FPR_0.01_filtered.h5\n",
    "path_cellbender = '../../data/CellBenderV2'\n",
    "\n",
    "# Path to scR-Invex output\n",
    "# Tool at https://github.com/getzlab/scrinvex\n",
    "path_scrinvex = '../../data/scrinvex'\n",
    "\n",
    "# Path to aggregation output\n",
    "path_aggregation = '../../data/aggregation'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine CellRanger and CellBender and Calculate some basic QC metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Samples to aggregate\n",
    "sample_list = ['1364', '1452', '1579', '1650', '1690', '1693', '1703', '1716', \n",
    "               '1733', '1739', '1763', '1773', '1785', '1800', '1801']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual</th>\n",
       "      <th>disease</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1364</td>\n",
       "      <td>ICM</td>\n",
       "      <td>F</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1452</td>\n",
       "      <td>NF</td>\n",
       "      <td>F</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1579</td>\n",
       "      <td>ICM</td>\n",
       "      <td>F</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1650</td>\n",
       "      <td>NF</td>\n",
       "      <td>M</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1690</td>\n",
       "      <td>NF</td>\n",
       "      <td>F</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1693</td>\n",
       "      <td>ICM</td>\n",
       "      <td>M</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1703</td>\n",
       "      <td>ICM</td>\n",
       "      <td>M</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1716</td>\n",
       "      <td>NF</td>\n",
       "      <td>F</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1733</td>\n",
       "      <td>ICM</td>\n",
       "      <td>F</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1739</td>\n",
       "      <td>NF</td>\n",
       "      <td>M</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1763</td>\n",
       "      <td>NF</td>\n",
       "      <td>M</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1773</td>\n",
       "      <td>ICM</td>\n",
       "      <td>M</td>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1785</td>\n",
       "      <td>NF</td>\n",
       "      <td>F</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1800</td>\n",
       "      <td>ICM</td>\n",
       "      <td>F</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1801</td>\n",
       "      <td>NF</td>\n",
       "      <td>M</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   individual disease sex  age\n",
       "9        1364     ICM   F   55\n",
       "1        1452      NF   F   47\n",
       "2        1579     ICM   F   47\n",
       "14       1650      NF   M   58\n",
       "3        1690      NF   F   63\n",
       "4        1693     ICM   M   64\n",
       "12       1703     ICM   M   55\n",
       "11       1716      NF   F   64\n",
       "8        1733     ICM   F   64\n",
       "10       1739      NF   M   52\n",
       "13       1763      NF   M   58\n",
       "5        1773     ICM   M   59\n",
       "7        1785      NF   F   53\n",
       "0        1800     ICM   F   62\n",
       "6        1801      NF   M   42"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Phenotype values for age, sex, and disease status\n",
    "sex_map = {'1364':'F',\n",
    "           '1452':'F',\n",
    "           '1579':'F',\n",
    "           '1650':'M',\n",
    "           '1690':'F',\n",
    "           '1693':'M',\n",
    "           '1703':'M',\n",
    "           '1716':'F',\n",
    "           '1733':'F',\n",
    "           '1739':'M',\n",
    "           '1763':'M',\n",
    "           '1773':'M',\n",
    "           '1785':'F',\n",
    "           '1800':'F',\n",
    "           '1801':'M'}\n",
    "disease_map = {'1800':'ICM',\n",
    "               '1452':'NF',\n",
    "               '1579':'ICM',\n",
    "               '1690':'NF',\n",
    "               '1693':'ICM',\n",
    "               '1773':'ICM',\n",
    "               '1801':'NF',\n",
    "               '1785':'NF',\n",
    "               '1733':'ICM',\n",
    "               '1364':'ICM',\n",
    "               '1739':'NF',\n",
    "               '1716':'NF',\n",
    "               '1703':'ICM',\n",
    "               '1763':'NF',\n",
    "               '1650':'NF'}\n",
    "age_map = {'1364':55,\n",
    "           '1452':47,\n",
    "           '1579':47,\n",
    "           '1650':58,\n",
    "           '1690':63,\n",
    "           '1693':64,\n",
    "           '1703':55,\n",
    "           '1716':64,\n",
    "           '1733':64,\n",
    "           '1739':52,\n",
    "           '1763':58,\n",
    "           '1773':59,\n",
    "           '1785':53,\n",
    "           '1800':62,\n",
    "           '1801':42}\n",
    "mydicts = [disease_map, sex_map, age_map]\n",
    "phenos = pd.concat([pd.Series(d) for d in mydicts], axis=1).fillna(0).reset_index().sort_values('index')\n",
    "phenos.columns = ['individual','disease','sex','age']\n",
    "phenos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Function to process each sample\n",
    "Adds the following information/QC metrics to the anndata:\n",
    "1. Phenotypic information from above\n",
    "2. ncount: Total UMI per nucleus based on CellBender and CellRanger\n",
    "3. ngenes: Total unique genes per nucleus based on CellBender and CellRanger\n",
    "4. percent_mito: Percent mitochondrial mapped reads, CellRanger only\n",
    "5. exon_prop: Proportion of reads mapping exclusively to exons\n",
    "6. entropy: Nucleus-based entropy based on CellBender and Cellranger\n",
    "'''\n",
    "def import_sample(sample,\n",
    "                  path_cellranger,\n",
    "                  path_cellbender,\n",
    "                  path_scrinvex,\n",
    "                  phenos):\n",
    "    # CellRanger Counts -- all barcodes (not just CellRanger derived cells)\n",
    "    adata_cr = sc.read_10x_h5(os.path.join(path_cellranger, sample + '.raw_feature_bc_matrix.h5'))\n",
    "    adata_cr.var_names_make_unique()\n",
    "    \n",
    "    # CellBender Cells and Counts -- we will use the non-empty droplets from CellBender as our cells\n",
    "    adata_cb = sc.read_10x_h5(os.path.join(path_cellbender, sample + '_out_FPR_0.01_filtered.h5'))\n",
    "    adata_cb.var_names_make_unique()\n",
    "    \n",
    "    # Add the cellranger counts to layers\n",
    "    adata_cr = adata_cr[adata_cb.obs.index.tolist(),:,] # subset raw to cellbender cells and order appropriately\n",
    "    adata_cb.layers['cellranger_raw'] = adata_cr.X # keep the cellranger counts in the object\n",
    "    \n",
    "    # Add in some sample based annotations: In this case sample & individual are the same\n",
    "    # But this data will eventually be combined with the DCM/HCM data in which case this isn't true\n",
    "    # so best to keep all the info for now\n",
    "    adata_cb.obs['sample'] = sample\n",
    "    adata_cb.obs['individual'] = sample\n",
    "    for phen in [y for y in phenos.columns if not y == 'individual']:\n",
    "        adata_cb.obs[phen] = phenos[phenos['individual'] == sample][phen].iloc[0]\n",
    "\n",
    "    # QC metrics we need to calculate\n",
    "    # nUMI and nGenes per nucleus\n",
    "    adata_cb.obs['cellranger_ncount'] = np.array(adata_cb.layers['cellranger_raw'].sum(1)).squeeze()\n",
    "    adata_cb.obs['cellbender_ncount'] = np.array(adata_cb.X.sum(1)).squeeze()\n",
    "    adata_cb.obs['cellranger_ngenes'] = np.array((adata_cb.layers['cellranger_raw'] > 0).sum(1)).squeeze()\n",
    "    adata_cb.obs['cellbender_ngenes'] = np.array((adata_cb.X > 0).sum(1)).squeeze()\n",
    "    \n",
    "    # % MT reads\n",
    "    mito_genes = adata_cb.var_names.str.startswith('MT-') # get MT genes\n",
    "    adata_cb.obs['cellranger_percent_mito'] = np.sum(adata_cb[:, mito_genes].layers['cellranger_raw'], axis=1).A1 / np.sum(adata_cb.layers['cellranger_raw'], axis=1).A1\n",
    "    \n",
    "    # Exon proportion\n",
    "    scrinvex = pd.read_csv(os.path.join(path_scrinvex,sample+'.scrinvex.collapsed.tsv'),sep='\\t')\n",
    "    scrinvex = scrinvex[scrinvex['barcode'].isin(adata_cb.obs.index)]\n",
    "    scrinvex['exon_prop'] = scrinvex['exons'] / scrinvex[['introns','junctions','exons']].sum(1)\n",
    "    scrinvex_lookup = {x:y for x,y in zip(scrinvex['barcode'],scrinvex['exon_prop'])}\n",
    "    adata_cb.obs['exon_prop'] = [scrinvex_lookup[x] for x in adata_cb.obs.index]\n",
    "    \n",
    "    # Remove cells with empty cellbender counts -- these will fail and need to be discarded\n",
    "    # This happens when CellBender removes all the counts for a droplet,\n",
    "    # and normally indicates that something is off with that barcode\n",
    "    adata_cb = adata_cb[adata_cb.obs['cellbender_ncount'] > 0].copy()\n",
    "\n",
    "    # Calculate and append entropy\n",
    "    entropy_cb = np.zeros(adata_cb.X.shape[0])\n",
    "    for i in range(adata_cb.X.shape[0]):\n",
    "        # calculate entropy (based on cellbender)\n",
    "        entropy_cb[i] = ndd.entropy(np.array(adata_cb.X[i, :].todense()).squeeze(), k=adata_cb.X.shape[1])\n",
    "    adata_cb.obs['cellbender_entropy'] = entropy_cb\n",
    "    entropy_cr = np.zeros(adata_cb.X.shape[0])\n",
    "    for i in range(adata_cb.X.shape[0]):\n",
    "        # calculate entropy (based on cellranger)\n",
    "        entropy_cr[i] = ndd.entropy(np.array(adata_cb.layers['cellranger_raw'][i, :].todense()).squeeze(), k=adata_cb.X.shape[1])\n",
    "    adata_cb.obs['cellranger_entropy'] = entropy_cr\n",
    "    return(adata_cb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Build the adata object for each individual sample and save this adata for subsequent aggregation\n",
    "for s in sample_list:\n",
    "    print(s)\n",
    "    adata = import_sample(s,\n",
    "                          path_cellranger,\n",
    "                          path_cellbender,\n",
    "                          path_scrinvex,\n",
    "                          phenos)\n",
    "    adata.write(os.path.join(path_aggregation,s+'.preqc.h5ad'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Append Scrublet Scores to Nuclei from Each Sample\n",
    "This is based on both CellBender and CellRanger separately."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate Scrublet Scores for Each Nucleus in each sample\n",
    "# Based on CellRanger and CellBender\n",
    "# Save the output to an h5ad file\n",
    "for s in sample_list:\n",
    "    print(s)\n",
    "    print('CellBender...')\n",
    "    adata = sc.read(os.path.join(path_aggregation, s + '.preqc.h5ad'))\n",
    "    scrub = scr.Scrublet(adata.X)\n",
    "    doublet_scores, predicted_doublets = scrub.scrub_doublets()\n",
    "    adata.obs['cellbender_doublet_scores'] = doublet_scores\n",
    "    adata.obs['cellbender_predicted_doublets'] = predicted_doublets\n",
    "    \n",
    "    print('CellRanger...')\n",
    "    scrub = scr.Scrublet(adata.layers['cellranger_raw'])\n",
    "    doublet_scores, predicted_doublets = scrub.scrub_doublets()\n",
    "    adata.obs['cellranger_doublet_scores'] = doublet_scores\n",
    "    adata.obs['cellranger_predicted_doublets'] = predicted_doublets\n",
    "    adata.write(os.path.join(path_aggregation, s + '.preqc.scrub.h5ad'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove the pre-scrublet data object as we don't need it anymore and it is just taking up space\n",
    "for s in sample_list:\n",
    "    os.system('rm ' + os.path.join(path_aggregation,s+'.preqc.h5ad'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
